### PR TITLE
pacific: doc/crimson: document wait states.

### PR DESCRIPTION
**Pull request by rzarzynski**
Signed-off-by: Radoslaw Zarzynski <rzarzyns@redhat.com>
(cherry picked from commit 7110bbed121f025b10a4f2a74a235316f79194bb)

## Checklist
- [ ] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

### CODE REVIEW

**Comment by rzarzynski** [comment]
The `master` PR: #40282.

**Comment by tchaikov** [comment]
@rzarzynski why would you want to backport this document PR? i think we backport bug fixes to LTS branches in general.

**Comment by neha-ojha** [comment]
> > why would you want to backport this document PR? i think we backport bug fixes to LTS branches in general.
> 
> The benefit I see is having the doc easier to find (basically it would be present in two places). However, I admit this is a little benefit, so dropping the backport would be fine. On the other hand, the costs (risks) of merging is even lower as it's docs-related solely.

The plan is the make readthedocs default to the most recent stable version(very soon pacific) instead of master. That would require us to backport doc PRs to pacific, so this is a good start.

**Review by neha-ojha — APPROVED** [review]

**Comment by tchaikov** [comment]
I'd encourage developers to work on master and read document generated from master branch.

**Comment by tchaikov** [comment]
> 
> The plan is the make readthedocs default to the most recent stable version(very soon pacific) instead of master. That would require us to backport doc PRs to pacific, so this is a good start.

I assume you've read this document thoroughly as you've approved it. but please note allow me clarify, the target audience of this document is not users who expect stable version, but the developers.

File: doc/dev/crimson/pipeline.rst [hunk 1]
```diff
@@ -0,0 +1,95 @@
...
+
+
+  ``ClientRequest::ConnectionPipeline::await_map``
+    wait for particular OSDMap version is available at the OSD level
+  ``ClientRequest::ConnectionPipeline::get_pg``
+    wait a particular PG becomes available on OSD
+  ``ClientRequest::PGPipeline::await_map``
+    wait on a PG being advanced to particular epoch
+  ``ClientRequest::PGPipeline::wait_for_active``
+    wait on a PG becomes ``is_active()``
```
Comment (ronen-fr) [inline doc/dev/crimson/pipeline.rst:27]: Maybe:
"wait for a PG to become 'active' (i.e. have is_active() asserted)"

File: doc/dev/crimson/pipeline.rst [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+  ``ClientRequest::PGPipeline::recover_missing``
+    wait on an object has been recovered
```
Comment (ronen-fr) [inline doc/dev/crimson/pipeline.rst:29]: Maybe:
"Wait on an object to be recovered"

(also - I'd add a comment mentioning how that will come to pass)

File: doc/dev/crimson/pipeline.rst [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+  ``ClientRequest::PGPipeline::get_obc``
+    wait on an object context becomes locked
```
Comment (ronen-fr) [inline doc/dev/crimson/pipeline.rst:31]: Similarly:
"to be available for locking. The obc will be locked before this OP is allowed to continue"

File: doc/dev/crimson/pipeline.rst [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+  ``ClientRequest::PGPipeline::process``
+    wait if any other ``MOSDOp`` message is handled against this PG
+
+At any moment, a ``ClientRequest`` being served should be in one and only one
+of these  phases. Similarly, an object denoting particular phase can host not
```
Comment (ronen-fr) [inline doc/dev/crimson/pipeline.rst:36]: Note: 'phases' is unclear at this point. The list above is titled as "here is where you may wait", but
it isn't clear that we treat these as a list of phases.

File: doc/dev/crimson/pipeline.rst [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+freeing it for another ``ClientRequest`` instance. All these phases form a pipeline
+which assures the order is preserved.
+
+These pipeline phases are divided into two ordering domains: ``ConnectionPipeline``
+and ``PGPipeline``. The former ensures order across a client connection while
+the latter does that across a PG. That is, requests originating from the same
+connection are executed in the same order as they were sent by the client.
+The same applies to the PG domain: when requests from multiple connections reach
+a PG, they are executed in the same order as they entered a first blocking phase
+of the ``PGPipeline``.
```
Comment (ronen-fr) [inline doc/dev/crimson/pipeline.rst:51]: Please help clarify: how are requests divided (moved?) between the domains? what is the relationship between
requests in the separate groups?

**Comment by rzarzynski** [comment]
@ronen-fr: I'm going to address the comments in a follow-up for `master`. Thanks for the review!

File: doc/dev/crimson/pipeline.rst (hunk 5 above)
Comment (rzarzynski) [reply doc/dev/crimson/pipeline.rst:51]: I believe there is no any special moving. The same `handle.enter()` mechanism is used for the crossing.

**Review by tchaikov — CHANGES_REQUESTED** [review]
I don't agree on backporting dev docs to stable branches.

**Pull request closed without merging**